src/Block/Fetch.hs

```diff
@@ -1,108 +1,196 @@
-module Block.Fetch (wsApp) where
+module Block.Fetch (
+    OgmiosWorkerMVar (MkOgmiosWorkerMVar),
+    OgmiosInfo (..),
+    StartBlockFetcherError (..),
+    StopBlockFetcherError (..),
+    startBlockFetcher,
+    stopBlockFetcher,
+    isBlockFetcherRunning,
+    createStoppedFetcher,
+) where
 
-import Colog (logError, logInfo, logWarning)
-import Control.Exception (Exception)
-import Control.Monad (forever, unless)
-import Control.Monad.Reader (ask)
+import Control.Concurrent.MVar (MVar, isEmptyMVar, newEmptyMVar, tryPutMVar, tryTakeMVar)
+import Control.Exception (Exception, onException)
+import Control.Monad (forever, unless, void)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Control.Monad.Logger (MonadLogger, logErrorNS, logInfoNS, logWarnNS, runStdoutLoggingT)
+import Control.Monad.Reader qualified as Reader
+import Control.Monad.Reader.Has (Has, MonadReader, ask, runReaderT)
 import Control.Monad.Trans (liftIO)
 import Data.Aeson qualified as Json
 import Data.ByteString.Base64 qualified as BSBase64
 import Data.Map qualified as Map
-import Data.Set qualified as Set
 import Data.Text (Text)
 import Data.Text qualified as Text
 import Data.Text.Encoding qualified as Text
+import Hasql.Connection qualified as Hasql
 import Hasql.Session qualified as Session
 import Network.WebSockets qualified as WS
+import UnliftIO.Async (Async)
 import UnliftIO.Async qualified as Async
 import UnliftIO.Concurrent (threadDelay)
 
-import App (App)
-import App.Env (Env (..))
-import App.FirstFetchBlock
-import App.RequestedDatumHashes qualified as RequestedDatumHashes
+import Api.Types (FirstFetchBlock (FirstFetchBlock))
+import Block.Filter (DatumFilter, runDatumFilter)
 import Block.Types (AlonzoBlock (..), AlonzoTransaction (..), Block (..), FindIntersectResult (..), OgmiosFindIntersectResponse, OgmiosRequestNextResponse, OgmiosResponse (..), RequestNextResult (..), mkFindIntersectRequest, mkRequestNextRequest)
```

Comment:
Really long line here (I know it's not part of the changes in this PR, but still :))